### PR TITLE
Use Pulumi random provider

### LINKED ISSUE #1628: Pulumi: secret management
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

Replace current management of Pulumi secrets

## :steam_locomotive: How could this be done?

Instead of generating these in our code and providing them as `Pulumi.config` secure values, we should use the Pulumi random provider.

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).
- [x] You have formatted your code using appropriate automated tools (for example `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>` for Powershell).

### :arrow_heading_up: Summary

Replace custom secret generation with Pulumi random provider.

Note that the following secrets were not replaced:
- [SHM `password-domain-ldap-searcher`](#1645) (as this is needed by both SHM and SRE and replacing it would involve a rewrite of how we pass secrets between stacks)
- SHM `verification-azuread-custom-domain` (could possibly be passed as an argument to the deployment code in some way)
- [SRE `password-domain-ldap-searcher`](#1645) (copied from SHM stack)
- [SRE `salt-dns-server-admin`](#1644) (this cannot trivially be generated using Pulumi's random provider)

### :closed_umbrella: Related issues

Closes #1628 

### :microscope: Tests

~Not tested due to #1641.~ Tested on SHM green and SREs leonardo and donatello.

### CODE REVIEW

**Comment by JimMadge** [comment]
This looks like the right idea to me.

The provider has a lot of useful features, the password resources are cryptographically secure.
I think using (and contributing) to this is likely to be better than what we make ourselves.

File: data_safe_haven/commands/deploy_shm.py [hunk 1]
```diff
@@ -50,17 +50,7 @@ def deploy_shm(
...
         stack.add_option("azure-native:tenantId", config.azure.tenant_id, replace=False)
         # Add necessary secrets
-        stack.add_secret("password-domain-admin", password(20), replace=False)
-        stack.add_secret(
-            "password-domain-azure-ad-connect", password(20), replace=False
-        )
-        stack.add_secret(
-            "password-domain-computer-manager", password(20), replace=False
-        )
         stack.add_secret("password-domain-ldap-searcher", password(20), replace=False)
```
Comment (JimMadge) [inline data_safe_haven/commands/deploy_shm.py:53]: Is this a credential Guacamole users to find users?
Are we not able to replace it here?

**Review by JimMadge — COMMENTED** [review]
Looks good. Definitely a move in the right direction.

It would be good to remove the need for our own password/secret generation and handling all together.

I see that you've already written about this in the PR. Perhaps we can open issues for each/all of these and give an impression of how much effort they are?
No need to block this.

**Comment by JimMadge** [comment]
> SRE salt-dns-server-admin (this cannot trivially be generated using Pulumi's random provider)

Why is that?

File: data_safe_haven/commands/deploy_shm.py (hunk 1 above)
Comment (jemrobinson) [reply data_safe_haven/commands/deploy_shm.py:53]: It's the password for a user account on the domain controller, so both the SHM (to create it) and the SRE (to use it) need to know what it is.

**Comment by jemrobinson** [comment]
> > SRE salt-dns-server-admin (this cannot trivially be generated using Pulumi's random provider)
> 
> Why is that?

It needs to be a byte string of the form: `$prefix$rounds$output` where `prefix` is either `2a` or `2b`, `rounds` is a zero-padded integer from 4-31 inclusive and output is a 16 byte integer encoded into a base64 byte string using a non-standard version of base64. We currently use the `bcrypt` library to do this, but the `gensalt` function gets its own random number from `urandom` - it doesn't take a random number (which we could provide) as input.

So, we *could* use the Pulumi random provider but we'd need to write our own version of this function and keep it up to date with any future changes to the bcrypt protocol.

**Comment by JimMadge** [comment]
I think I remember looking at the salt before, making sure it was idempotent.

Taking a quick look, I think we are handling that in the most sensible way. Create the data, store it in configuration.

**Pull request merged**